Rewrite connection string handling

adamralph opened the pull request:
fixes #298

adamralph commented on src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs:
```diff
@@ -62,14 +149,51 @@ void SetDefaultClientProperties(string endpointName)
+            {
+                var message = invalidOptionsMessage.ToString().TrimEnd('\r', '\n');
+
+                Logger.Error(message);
```
I'd love to get rid of this. We're throwing the exception up the stack so logging the message here feels inappropriate. By this logic we should log every exception message before we throw it. Doesn't make sense to me.

bording replied:
The exception will end up getting logged anyway, right? If so, then yes this separate logging call is redundant.

adamralph replied:
Exactly, or if not, it will go all the way up the stack and kill the process as unhandled. Regardless, logging it is not concern of this class. I'll add another commit to remove this.

bording commented on src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs:
```diff
@@ -2,58 +2,145 @@
+
+        public IReadOnlyDictionary<string, string> ClientProperties { get; }
+
+        private ConnectionConfiguration(
```
Drop the `private`.

bording commented on src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs:
```diff
@@ -2,58 +2,145 @@
-        public ConnectionConfiguration(string endpointName)
+        public string CertPassphrase { get; }
+
+        public IReadOnlyDictionary<string, string> ClientProperties { get; }
```
Given that this is an internal class, I'd prefer to drop the interface and the readonly, and keep this as a Dictionary instead.

adamralph replied:
Yeah, come to think of it, I agree. I'll send a fixup!

The commit that added this has the message "made ConnectionConfiguration immutable" but when I squash I'll alter the message, since it won't be immutable anymore. That said, I'm not sure if we want to keep the micro commits or not before merging.

bording commented on src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs:
```diff
@@ -2,58 +2,145 @@
+            var certPath = GetValue(dictionary, "certPath", defaultCertPath);
+            var certPassPhrase = GetValue(dictionary, "certPassphrase", defaultCertPassphrase);
+
+            var host = default(string);
```
Formatting: line break here please.

adamralph replied:
Done, but seems like GitHub doesn't realise it.

bording commented on src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs:
```diff
@@ -2,58 +2,145 @@
+
+                var parts = hostsAndPorts[0].Split(':');
+                host = parts.ElementAt(0);
+                if (host.Length == 0)
```
Formatting, line break before the `if` please

bording commented on src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs:
```diff
@@ -2,58 +2,145 @@
+                    invalidOptionsMessage.AppendLine("Empty host name in 'host' connection string option.");
+                }
+
+                if (parts.Length > 1 && !int.TryParse(parts[1], out port))
```
I'll need to study this section a bit more compared to the old `ParseHosts` method, but it feels like there might be a logic change here aside from the change to using `TryParse`.

bording replied:
Spent more time looking at this, and this looks fine. In fact, it looks like it handles even more parsing scenarios than the previous code did, so 👍.

bording commented on src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs:
```diff
@@ -62,14 +149,51 @@ void SetDefaultClientProperties(string endpointName)
+                host, port, virtualHost, userName, password, requestedHeartbeat, retryDelay, useTls, certPath, certPassPhrase, clientProperties);
+        }
+
+        static string GetValue(IDictionary<string, string> dictionary, string key, string @default)
```
I'd prefer to drop the interface here as well.

bording replied:
Could we change the name of the default parameter to `defaultValue` to get rid of needing the `@` syntax?

bording commented on src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs:
```diff
@@ -62,14 +149,51 @@ void SetDefaultClientProperties(string endpointName)
+            return dictionary.TryGetValue(key, out value) ? value : @default;
         }
+
+        static T GetValue<T>(IDictionary<string, string> dictionary, string key, Convert<T> convert, T @default, StringBuilder invalidOptionsMessage)
```
I'd prefer to drop the interface here as well.

bording replied:
Same comment about the parameter here.

adamralph commented:
@bording the series of fixup!s should address all your comments.

adamralph commented on src/NServiceBus.RabbitMQ/Configuration/ConnectionConfiguration.cs:
```diff
@@ -2,58 +2,143 @@
+
+            var useTls = GetValue(dictionary, "useTls", bool.TryParse, defaultUseTls, invalidOptionsMessage);
+            var port = GetValue(dictionary, "port", int.TryParse, useTls ? defaultTlsPort : defaultPort, invalidOptionsMessage);
+            var virtualHost = GetValue(dictionary, "virtualHost", "/");
```
Oops, no, just missed them in my excitement. Will fix.

adamralph commented:
@bording I pushed a fixup! to add the missing constants.

bording commented:
@adamralph I think this is looking good. 

Given that we're still discussing #295 / #297 do you want to base this on develop instead and go ahead and get it in, or wait until we have decided on how to handle the virtual host stuff?

adamralph commented:
@bording there is indeed no need for this branch to be based on the branch for #297 so I've rebased this branch on to develop and squashed the fixup! commits. The rebase was far from trivial so you might want to have another glance at the diff.

bording commented:
@adamralph Do you want to squash this any further, or would your prefer to merge with all the intermediate commits?

adamralph commented:
@bording I'm not really sure. Each commit is self contained (i.e. none of them are effectively fixup!s for previous commits) but they are very granular. One thing we could do is squash the refactoring commits together but keep the test/functional changes separate like so:

```
refactor: merge ConnectionStringParser into ConnectionConfiguration
green: test that host port takes precedence over separate port
refactor: move connection string validation to before configuration construction
red: report all invalid values in a connection string
green: report all invalid values in a connection string
refactor: ConnectionConfiguration internals
remove redundant logging
```
